src/nbl/video/CVulkanSwapchain.cpp

```diff
@@ -14,6 +14,70 @@ CVulkanSwapchain::~CVulkanSwapchain()
+        vkPresentMode = VK_PRESENT_MODE_IMMEDIATE_KHR;
+    else if ((params.presentMode & ISurface::E_PRESENT_MODE::EPM_MAILBOX) == ISurface::E_PRESENT_MODE::EPM_MAILBOX)
+        vkPresentMode = VK_PRESENT_MODE_MAILBOX_KHR;
+    else if ((params.presentMode & ISurface::E_PRESENT_MODE::EPM_FIFO) == ISurface::E_PRESENT_MODE::EPM_FIFO)
+        vkPresentMode = VK_PRESENT_MODE_FIFO_KHR;
+    else if ((params.presentMode & ISurface::E_PRESENT_MODE::EPM_FIFO_RELAXED) == ISurface::E_PRESENT_MODE::EPM_FIFO_RELAXED)
+        vkPresentMode = VK_PRESENT_MODE_FIFO_RELAXED_KHR;
+
+    VkSwapchainCreateInfoKHR vk_createInfo = { VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR };
+    vk_createInfo.surface = vk_surface;
+    vk_createInfo.minImageCount = params.minImageCount;
+    vk_createInfo.imageFormat = getVkFormatFromFormat(params.surfaceFormat.format);
+    vk_createInfo.imageColorSpace = getVkColorSpaceKHRFromColorSpace(params.surfaceFormat.colorSpace);
+    vk_createInfo.imageExtent = { params.width, params.height };
+    vk_createInfo.imageArrayLayers = params.arrayLayers;
+    vk_createInfo.imageUsage = static_cast<VkImageUsageFlags>(params.imageUsage.value);
+    vk_createInfo.imageSharingMode = static_cast<VkSharingMode>(params.imageSharingMode);
+    vk_createInfo.queueFamilyIndexCount = params.queueFamilyIndexCount;
+    vk_createInfo.pQueueFamilyIndices = params.queueFamilyIndices;
+    vk_createInfo.preTransform = static_cast<VkSurfaceTransformFlagBitsKHR>(params.preTransform);
+    vk_createInfo.compositeAlpha = static_cast<VkCompositeAlphaFlagBitsKHR>(params.compositeAlpha);
+    vk_createInfo.presentMode = vkPresentMode;
+    vk_createInfo.clipped = VK_FALSE;
+    vk_createInfo.oldSwapchain = VK_NULL_HANDLE;
+    if (params.oldSwapchain && (params.oldSwapchain->getAPIType() == EAT_VULKAN))
+        vk_createInfo.oldSwapchain = IBackendObject::device_compatibility_cast<CVulkanSwapchain*>(params.oldSwapchain.get(), device.get())->getInternalObject();
+
+    VkSwapchainKHR vk_swapchain;
+    if (device->getFunctionTable()->vk.vkCreateSwapchainKHR(device->getInternalObject(), &vk_createInfo, nullptr, &vk_swapchain) != VK_SUCCESS)
+        return nullptr;
+
+    uint32_t imageCount;
+    VkResult retval = device->getFunctionTable()->vk.vkGetSwapchainImagesKHR(device->getInternalObject(), vk_swapchain, &imageCount, nullptr);
+    if ((retval != VK_SUCCESS) && (retval != VK_INCOMPLETE))
+        return nullptr;
+
+    assert(imageCount <= MAX_SWAPCHAIN_IMAGE_COUNT);
+
+    IGPUImage::SCreationParams imgParams;
+    imgParams.flags = static_cast<IGPUImage::E_CREATE_FLAGS>(0); // Todo(achal)
+    imgParams.type = CVulkanImage::ET_2D;
+    imgParams.format = params.surfaceFormat.format;
+    imgParams.extent = { params.width, params.height, 1u };
+    imgParams.mipLevels = 1u;
+    imgParams.arrayLayers = params.arrayLayers;
+    imgParams.samples = CVulkanImage::ESCF_1_BIT; // Todo(achal)
+    imgParams.usage = params.imageUsage;
```

Comment:
you want to retrieve VKImages here and create the `CVulkanImages` before you enter the swapchain constructor